File: supervisor/updater.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"syscall"
	"time"
)

const supervisorReleaseURL = "https://api.github.com/repos/andriyshevchenko/remote-copilot-mcp/releases/tags/supervisor-latest"

var notifyUpdaterOperator = NotifyOperator

// Updater checks the npm registry for new versions and performs updates.
type Updater struct {
	cfg     Config
	mcp     *MCPClient
	log     *Logger
	state   *UpdateStateStore
	startAt time.Time
	cancel  context.CancelFunc
	done    chan struct{}
}

func NewUpdater(cfg Config, mcp *MCPClient, log *Logger) *Updater {
	return &Updater{
		cfg:     cfg,
		mcp:     mcp,
		log:     log,
		state:   NewUpdateStateStore(cfg.Paths.UpdateState, log),
		startAt: time.Now(),
		done:    make(chan struct{}),
	}
}

// Start begins the update check loop.
func (u *Updater) Start() {
	ctx, cancel := context.WithCancel(context.Background())
	u.cancel = cancel
	go u.run(ctx)
}

// Stop signals the updater to shut down and waits.
func (u *Updater) Stop() {
	if u.cancel != nil {
		u.cancel()
	}
	<-u.done
}

func (u *Updater) run(ctx context.Context) {
	defer close(u.done)
	u.log.Info("Updater started (mode=%s)", u.cfg.Mode)

	// In development mode, check every PollInterval.
	// In production, check once per day at PollAtHour.
	for {
		var sleepDuration time.Duration
		if u.cfg.Mode == "development" {
			sleepDuration = u.cfg.PollInterval
		} else {
			sleepDuration = u.timeUntilNextPoll()
		}
		u.log.Debug("Updater: next version check in %v", sleepDuration.Round(time.Second))

		select {
		case <-ctx.Done():
			return
		case <-time.After(sleepDuration):
		}

		if ctx.Err() != nil {
			return
		}

		u.checkSupervisorUpdate(ctx)
	}
}

func (u *Updater) timeUntilNextPoll() time.Duration {
	now := time.Now()
	next := time.Date(now.Year(), now.Month(), now.Day(), u.cfg.PollAtHour, 0, 0, 0, now.Location())
	if next.Before(now) {
		next = next.Add(24 * time.Hour)
	}
	return time.Until(next)
}

type githubRelease struct {
	TagName string `json:"tag_name"`
	Name    string `json:"name"`
	Assets  []struct {
		Name string `json:"name"`
		URL  string `json:"browser_download_url"`
		Size int64  `json:"size"`
	} `json:"assets"`
}

func (u *Updater) getSupervisorRelease(ctx context.Context) (string, string, error) {
	ctx2, cancel := context.WithTimeout(ctx, 20*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx2, http.MethodGet, supervisorReleaseURL, nil)
	if err != nil {
		return "", "", err
	}
	req.Header.Set("Accept", "application/vnd.github+json")
	req.Header.Set("User-Agent", "sensorium-supervisor-updater")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", "", fmt.Errorf("GitHub releases HTTP %d", resp.StatusCode)
	}

	var release githubRelease
	if err := json.NewDecoder(resp.Body).Decode(&release); err != nil {
		return "", "", err
	}

	assetName := supervisorAssetName()
	for _, asset := range release.Assets {
		if asset.Name != assetName {
			continue
		}

		version := strings.TrimSpace(release.Name)
		if version == "" {
			version = strings.TrimSpace(release.TagName)
		}
		if version == "" {
			return "", "", fmt.Errorf("release version missing for %s", assetName)
		}
		if strings.TrimSpace(asset.URL) == "" {
			return "", "", fmt.Errorf("release asset URL missing for %s", assetName)
		}

		return version, asset.URL, nil
	}

	return "", "", fmt.Errorf("release asset %q not found", assetName)
}

func supervisorAssetName() string {
	suffix := ""
	if runtime.GOOS == "windows" {
		suffix = ".exe"
	}
	return fmt.Sprintf("sensorium-supervisor-%s-%s%s", runtime.GOOS, runtime.GOARCH, suffix)
}

func (u *Updater) getLocalSupervisorVersion() string {
	data, err := os.ReadFile(u.cfg.Paths.SupervisorVersion)
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(data))
}

func (u *Updater) setLocalSupervisorVersion(v string) {
	os.MkdirAll(u.cfg.DataDir, 0755)
	if err := atomicWrite(u.cfg.Paths.SupervisorVersion, []byte(v)); err != nil {
		u.log.Warn("Failed to write supervisor version file: %v", err)
	}
}

func (u *Updater) stagePendingSupervisorVersion(v string) error {
	if err := os.MkdirAll(filepath.Dir(u.cfg.Paths.PendingVersion), 0755); err != nil {
		return fmt.Errorf("create pending supervisor version dir: %w", err)
	}
	if err := atomicWrite(u.cfg.Paths.PendingVersion, []byte(v)); err != nil {
		return fmt.Errorf("write pending supervisor version: %w", err)
	}
	return nil
}

func (u *Updater) checkSupervisorUpdate(ctx context.Context) {
	uptime := time.Since(u.startAt)
	if uptime < u.cfg.MinUptime {
		u.log.Info("Deferring supervisor update — too early (uptime %v < %v)", uptime.Round(time.Second), u.cfg.MinUptime)
		return
	}

	remote, downloadURL, err := u.getSupervisorRelease(ctx)
	if err != nil {
		u.log.Warn("Failed to check supervisor release: %v", err)
		return
	}

	local := u.getLocalSupervisorVersion()
	if local == "" {
		u.log.Info("No local supervisor version recorded — storing %s", remote)
		u.setLocalSupervisorVersion(remote)
		return
	}

	if local == remote {
		u.log.Debug("Supervisor updater: version %s is up to date", local)
		return
	}

	u.log.Info("Supervisor update available: %s → %s", local, remote)
	coordLock, ok := AcquireUpdateCoordinatorLock(u.cfg.Paths.UpdateApplyLock, updateScopeSupervisor, u.log)
	if !ok {
		u.log.Info("Deferring supervisor binary update %s → %s due to active update apply lock", local, remote)
		return
	}
	defer coordLock.Release()

	markFailed := func(err error) {
		u.state.Transition(updateScopeSupervisor, updatePhaseFailed, remote, local, err.Error())
	}

	notifyUpdaterOperator(u.cfg, u.log, fmt.Sprintf("⚙️ Supervisor: updating supervisor binary %s → %s. Grace period %v. Supervisor process will restart — MCP server unaffected.", local, remote, u.cfg.GracePeriod), 0)

	select {
	case <-ctx.Done():
		markFailed(ctx.Err())
		return
	case <-time.After(u.cfg.GracePeriod):
	}

	if err := u.downloadSupervisorBinary(ctx, downloadURL); err != nil {
		markFailed(err)
		u.log.Error("Supervisor binary download failed: %v", err)
		notifyUpdaterOperator(u.cfg, u.log, fmt.Sprintf("🔴 Supervisor: supervisor binary update to %s failed during download.", remote), 0)
		return
	}

	if err := u.stagePendingSupervisorVersion(remote); err != nil {
		_ = os.Remove(u.cfg.Paths.PendingBinary)
		markFailed(err)
		u.log.Error("Failed to stage supervisor version %s: %v", remote, err)
		notifyUpdaterOperator(u.cfg, u.log, fmt.Sprintf("🔴 Supervisor: supervisor binary update to %s failed during staging.", remote), 0)
		return
	}
	u.state.Transition(updateScopeSupervisor, updatePhaseStaged, remote, local, "")
	notifyUpdaterOperator(u.cfg, u.log, fmt.Sprintf("⚙️ Supervisor: supervisor binary %s downloaded. Restarting supervisor to apply update — MCP server will continue running.", remote), 0)

	// Reset start time so minimum uptime is re-enforced after restart
	u.startAt = time.Now()

	isService, err := isWindowsService()
	if err != nil {
		markFailed(err)
		u.log.Error("Failed to detect service mode for restart: %v", err)
		notifyUpdaterOperator(u.cfg, u.log, "🔴 Supervisor: supervisor binary downloaded but service detection failed.", 0)
		return
	}
	u.state.Transition(updateScopeSupervisor, updatePhaseRestarting, remote, local, "")

	if isService {
		if err := scheduleServiceRestartForUpdate(u.log); err != nil {
			markFailed(err)
			u.log.Error("Failed to schedule service restart: %v", err)
			notifyUpdaterOperator(u.cfg, u.log, "🔴 Supervisor: supervisor binary downloaded but service restart scheduling failed.", 0)
		}
		return
	}

	if err := requestSupervisorRestart(u.cfg, u.log); err != nil {
		markFailed(err)
		u.log.Error("Failed to signal supervisor for restart: %v", err)
		notifyUpdaterOperator(u.cfg, u.log, "🔴 Supervisor: supervisor binary downloaded but restart signal failed.", 0)
	}
}

func (u *Updater) downloadSupervisorBinary(ctx context.Context, downloadURL string) error {
	if err := os.MkdirAll(u.cfg.Paths.BinaryDir, 0755); err != nil {
		return fmt.Errorf("create binary dir: %w", err)
	}

	tmpPath := u.cfg.Paths.PendingBinary + ".download"
	defer os.Remove(tmpPath)

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, downloadURL, nil)
	if err != nil {
		return err
	}
	req.Header.Set("User-Agent", "sensorium-supervisor-updater")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("download HTTP %d", resp.StatusCode)
	}

	f, err := os.OpenFile(tmpPath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0755)
	if err != nil {
		return err
	}

	written, copyErr := io.Copy(f, resp.Body)
	closeErr := f.Close()
	if copyErr != nil {
		return copyErr
	}
	if closeErr != nil {
		return closeErr
	}
	if written <= 0 {
		return fmt.Errorf("downloaded empty binary")
	}

	if err := os.Remove(u.cfg.Paths.PendingBinary); err != nil && !os.IsNotExist(err) {
		return err
	}
	if err := os.Rename(tmpPath, u.cfg.Paths.PendingBinary); err != nil {
		return err
	}

	u.log.Info("Supervisor binary downloaded to %s (%d bytes)", u.cfg.Paths.PendingBinary, written)
	return nil
}

func signalSelf(sig os.Signal) error {
	proc, err := os.FindProcess(os.Getpid())
	if err != nil {
		return err
	}
	return proc.Signal(sig)
}

func requestSupervisorRestart(cfg Config, log *Logger) error {
	if runtime.GOOS != "windows" {
		return signalSelf(syscall.SIGTERM)
	}

	exePath, err := os.Executable()
	if err != nil {
		return fmt.Errorf("resolve executable path: %w", err)
	}

	if err := launchWindowsApplyHelper(cfg, exePath, false); err != nil {
		return fmt.Errorf("launch apply helper for restart: %w", err)
	}

	if log != nil {
		log.Info("Apply helper launched — exiting cleanly for binary swap")
	}

	go stopSupervisor()
	return nil
}

func (u *Updater) killServer() {
	u.log.Info("Updater: stopping current MCP server for update")
	pid, err := ReadPIDFile(u.cfg.Paths.ServerPID)
	if err != nil {
		u.log.Warn("Could not read server PID file: %v", err)
		KillByPort(u.cfg.MCPHttpPort, u.log)
		return
	}
	if err := KillProcessDirect(pid, u.log); err != nil {
		u.log.Error("Failed to kill server (PID %d): %v", pid, err)
		KillByPort(u.cfg.MCPHttpPort, u.log)
	}
}

File: supervisor-dotnet/Services/Updater.cs
using System.Net.Http.Json;
using System.Runtime.InteropServices;
using System.Text.Json;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using Sensorium.Supervisor.Configuration;
using Sensorium.Supervisor.Infrastructure;

namespace Sensorium.Supervisor.Services;

public sealed class Updater : IUpdater
{
    private const string ReleaseUrl =
        "https://api.github.com/repos/andriyshevchenko/remote-copilot-mcp/releases/tags/supervisor-latest";

    private readonly SupervisorOptions _opts;
    private readonly IMcpClient _mcp;
    private readonly ITelegramNotifier _notify;
    private readonly ILogger<Updater> _log;
    private readonly UpdateStateStore _state;
    private readonly HttpClient _http;

    private readonly DateTime _startedAt = DateTime.UtcNow;
    private Task? _loop;
    private CancellationTokenSource? _cts;

    public Updater(
        IHttpClientFactory httpFactory,
        IOptions<SupervisorOptions> opts,
        IMcpClient mcp,
        ITelegramNotifier notify,
        ILogger<Updater> log)
    {
        _opts = opts.Value;
        _mcp = mcp;
        _notify = notify;
        _log = log;
        _http = httpFactory.CreateClient("github");
        _http.Timeout = TimeSpan.FromSeconds(30);
        _http.DefaultRequestHeaders.Add("User-Agent", "sensorium-supervisor-updater");
        _http.DefaultRequestHeaders.Add("Accept", "application/vnd.github+json");
        _state = new UpdateStateStore(_opts.Paths.UpdateState, log);
    }

    public void Start(CancellationToken ct)
    {
        _cts = CancellationTokenSource.CreateLinkedTokenSource(ct);
        _loop = RunLoopAsync(_cts.Token);
        _log.LogInformation("Updater started (mode={Mode})", _opts.Mode);
    }

    public async Task StopAsync()
    {
        _cts?.Cancel();
        if (_loop != null)
        {
            try { await _loop.ConfigureAwait(false); }
            catch (OperationCanceledException) { /* expected */ }
        }
    }

    private async Task RunLoopAsync(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            TimeSpan delay = _opts.Mode == "development"
                ? _opts.PollInterval
                : TimeUntilNextPoll();

            _log.LogDebug("Updater: next check in {Delay}", delay);

            try { await Task.Delay(delay, ct).ConfigureAwait(false); }
            catch (OperationCanceledException) { return; }

            await CheckSupervisorUpdateAsync(ct).ConfigureAwait(false);
        }
    }

    private TimeSpan TimeUntilNextPoll()
    {
        var now = DateTime.Now;
        var next = new DateTime(now.Year, now.Month, now.Day, _opts.PollAtHour, 0, 0, DateTimeKind.Local);
        if (next <= now) next = next.AddDays(1);
        return next - now;
    }

    private async Task CheckSupervisorUpdateAsync(CancellationToken ct)
    {
        var uptime = DateTime.UtcNow - _startedAt;
        if (uptime < _opts.MinUptime)
        {
            _log.LogInformation("Deferring update check — too early (uptime {Uptime} < {Min})",
                uptime.ToString(@"hh\:mm\:ss"), _opts.MinUptime);
            return;
        }

        string remoteVersion, downloadUrl;
        try
        {
            (remoteVersion, downloadUrl) = await GetLatestReleaseAsync(ct).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            _log.LogWarning(ex, "Failed to check supervisor release");
            return;
        }

        string localVersion = ReadLocalVersion();
        if (string.IsNullOrEmpty(localVersion))
        {
            _log.LogInformation("No local supervisor version recorded — storing {Version}", remoteVersion);
            WriteLocalVersion(remoteVersion);
            return;
        }

        if (localVersion == remoteVersion)
        {
            _log.LogDebug("Supervisor is up to date ({Version})", localVersion);
            return;
        }

        _log.LogInformation("Supervisor update available: {Local} → {Remote}", localVersion, remoteVersion);

        using var coordLock = UpdateCoordinatorLock.TryAcquire(_opts.Paths.UpdateApplyLock, "supervisor", _log);
        if (coordLock is null)
        {
            _log.LogInformation("Deferring update {Local} → {Remote}: coordinator lock busy", localVersion, remoteVersion);
            return;
        }

        await _notify.NotifyAsync(
            $"⚙️ Supervisor: updating binary {localVersion} → {remoteVersion}. " +
            $"Grace period {_opts.GracePeriod.TotalSeconds}s. Supervisor will restart — MCP unaffected.",
            ct: ct).ConfigureAwait(false);

        try { await Task.Delay(_opts.GracePeriod, ct).ConfigureAwait(false); }
        catch (OperationCanceledException)
        {
            _state.Transition("supervisor", UpdatePhase.Failed, remoteVersion, localVersion, "cancelled during grace period");
            return;
        }

        try
        {
            await DownloadBinaryAsync(downloadUrl, ct).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            _state.Transition("supervisor", UpdatePhase.Failed, remoteVersion, localVersion, ex.Message);
            _log.LogError(ex, "Supervisor binary download failed");
            await _notify.NotifyAsync($"🔴 Supervisor: update to {remoteVersion} failed during download.", ct: ct).ConfigureAwait(false);
            return;
        }

        try
        {
            await StagePendingVersionAsync(remoteVersion).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            TryDeleteFile(_opts.Paths.PendingBinary);
            _state.Transition("supervisor", UpdatePhase.Failed, remoteVersion, localVersion, ex.Message);
            _log.LogError(ex, "Failed to stage version {Version}", remoteVersion);
            await _notify.NotifyAsync($"🔴 Supervisor: update to {remoteVersion} failed during staging.", ct: ct).ConfigureAwait(false);
            return;
        }

        _state.Transition("supervisor", UpdatePhase.Staged, remoteVersion, localVersion);
        await _notify.NotifyAsync(
            $"⚙️ Supervisor: binary {remoteVersion} downloaded. Restarting supervisor — MCP will continue running.",
            ct: ct).ConfigureAwait(false);

        _state.Transition("supervisor", UpdatePhase.Restarting, remoteVersion, localVersion);
        RequestRestart();
    }

    private async Task<(string Version, string DownloadUrl)> GetLatestReleaseAsync(CancellationToken ct)
    {
        using var cts = CancellationTokenSource.CreateLinkedTokenSource(ct);
        cts.CancelAfter(TimeSpan.FromSeconds(20));

        using var resp = await _http.GetAsync(ReleaseUrl, cts.Token).ConfigureAwait(false);
        resp.EnsureSuccessStatusCode();

        using var doc = await JsonDocument.ParseAsync(
            await resp.Content.ReadAsStreamAsync(cts.Token).ConfigureAwait(false),
            cancellationToken: cts.Token).ConfigureAwait(false);

        string assetName = SupervisorAssetName();
        var root = doc.RootElement;

        string version = root.TryGetProperty("name", out var name) ? name.GetString()?.Trim() ?? "" : "";
        if (string.IsNullOrEmpty(version) && root.TryGetProperty("tag_name", out var tag))
            version = tag.GetString()?.Trim() ?? "";

        if (string.IsNullOrEmpty(version))
            throw new InvalidOperationException("Release version is empty");

        foreach (var asset in root.GetProperty("assets").EnumerateArray())
        {
            if (asset.GetProperty("name").GetString() == assetName)
            {
                string url = asset.GetProperty("browser_download_url").GetString()
                    ?? throw new InvalidOperationException("Asset download URL is empty");
                return (version, url);
            }
        }

        throw new InvalidOperationException($"Release asset '{assetName}' not found");
    }

    private static string SupervisorAssetName()
    {
        string os = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "windows"
            : RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? "darwin"
            : "linux";

        string arch = RuntimeInformation.ProcessArchitecture switch
        {
            Architecture.X64 => "amd64",
            Architecture.Arm64 => "arm64",
            _ => RuntimeInformation.ProcessArchitecture.ToString().ToLowerInvariant()
        };

        string ext = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? ".exe" : "";
        return $"sensorium-supervisor-{os}-{arch}{ext}";
    }

    private async Task DownloadBinaryAsync(string downloadUrl, CancellationToken ct)
    {
        Directory.CreateDirectory(_opts.Paths.BinaryDir);

        string tmpPath = _opts.Paths.PendingBinary + ".download";
        try
        {
            using var cts = CancellationTokenSource.CreateLinkedTokenSource(ct);
            cts.CancelAfter(TimeSpan.FromMinutes(5));

            using var resp = await _http.GetAsync(downloadUrl, HttpCompletionOption.ResponseHeadersRead, cts.Token)
                .ConfigureAwait(false);
            resp.EnsureSuccessStatusCode();

            await using var fs = new FileStream(tmpPath, FileMode.Create, FileAccess.Write, FileShare.None,
                bufferSize: 81920, useAsync: true);
            await resp.Content.CopyToAsync(fs, cts.Token).ConfigureAwait(false);

            if (fs.Length == 0)
                throw new InvalidOperationException("Downloaded empty binary");

            _log.LogInformation("Supervisor binary downloaded ({Bytes} bytes); staging to {Path}", fs.Length, _opts.Paths.PendingBinary);
        }
        catch
        {
            TryDeleteFile(tmpPath);
            throw;
        }

        TryDeleteFile(_opts.Paths.PendingBinary);
        File.Move(tmpPath, _opts.Paths.PendingBinary, overwrite: false);
    }

    private async Task StagePendingVersionAsync(string version)
    {
        Directory.CreateDirectory(Path.GetDirectoryName(_opts.Paths.PendingVersion)!);
        var tmp = _opts.Paths.PendingVersion + ".tmp." + System.Environment.ProcessId;
        await File.WriteAllTextAsync(tmp, version).ConfigureAwait(false);
        File.Move(tmp, _opts.Paths.PendingVersion, overwrite: true);
    }

    private void RequestRestart()
    {
        _log.LogInformation("Requesting supervisor restart for binary swap");
        // Signal the host to shut down — the apply helper will swap the binary and relaunch.
        // Run on a separate thread so we don't block the updater loop; log any failure.
        Task.Run(() =>
        {
            try { SupervisorShutdown.RequestShutdown(); }
            catch (Exception ex) { _log.LogError(ex, "Failed to request supervisor shutdown for restart"); }
        });
    }

    private string ReadLocalVersion()
    {
        try { return File.Exists(_opts.Paths.SupervisorVersion) ? File.ReadAllText(_opts.Paths.SupervisorVersion).Trim() : ""; }
        catch { return ""; }
    }

    private void WriteLocalVersion(string version)
    {
        try
        {
            Directory.CreateDirectory(_opts.DataDir);
            var tmp = _opts.Paths.SupervisorVersion + ".tmp." + System.Environment.ProcessId;
            File.WriteAllText(tmp, version);
            File.Move(tmp, _opts.Paths.SupervisorVersion, overwrite: true);
        }
        catch (Exception ex)
        {
            _log.LogWarning(ex, "Failed to write supervisor version file");
        }
    }

    private static void TryDeleteFile(string path)
    {
        try { File.Delete(path); } catch { /* best-effort */ }
    }
}

/// <summary>Hook used by the Updater to signal the host to shut down.</summary>
internal static class SupervisorShutdown
{
    private static IHostApplicationLifetime? _lifetime;

    public static void Register(IHostApplicationLifetime lifetime) => _lifetime = lifetime;

    public static void RequestShutdown() => _lifetime?.StopApplication();
}
